src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp

```diff
@@ -0,0 +1,901 @@
+{
+    std::lock_guard<std::recursive_mutex> lock(*getMutex());
+    bool all = true; // have all servers been discovered?
+
+    for (auto & svr : mp_builtin->m_DiscoveryServers)
+    {
+        all &= (svr.proxy != nullptr);
+
+        if (svr.proxy && !mp_EDP->areRemoteEndpointsMatched(svr.proxy))
+        {
+            queueParticipantForEDPMatch(svr.proxy);
+        }
+    }
+
+    return all;
+}
+
+void PDPServer::announceParticipantState(bool new_change, bool dispose /* = false */, WriteParams& )
+{
+    StatefulWriter * pW = dynamic_cast<StatefulWriter*>(mp_PDPWriter);
+    assert(pW);
+
+    // Servers only send direct DATA(p) to servers in order to allow discovery
+    if (new_change)
+    {
+        // only builtinprotocols uses new_change = true, delegate in base class
+        // in order to get the ParticipantProxyData into the WriterHistory and broadcast the first DATA(p)
+
+        WriteParams wp;
+        SampleIdentity local;
+        local.writer_guid(mp_PDPWriter->getGuid());
+        local.sequence_number(mp_PDPWriterHistory->next_sequence_number());
+        wp.sample_identity(local);
+        wp.related_sample_identity(local);
+
+        if (!dispose)
+        {
+            PDP::announceParticipantState(new_change, dispose, wp);
+        }
+        else
+        {   // we must assure when the server is dying that all client are send at least a DATA(p)
+            // note here we can no longer receive and DATA or ACKNACK from clients.
+            // In order to avoid that we send the message directly as in the standard stateless PDP
+            CacheChange_t* change = nullptr;
+
+            if ((change = pW->new_change([]() -> uint32_t {return DISCOVERY_PARTICIPANT_DATA_MAX_SIZE; },
+                NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key)))
+            {
+                // update the sequence number
+                change->sequenceNumber = mp_PDPWriterHistory->next_sequence_number();
+                change->write_params = wp;
+
+                std::lock_guard<std::recursive_timed_mutex> wlock(pW->getMutex());
+
+                RTPSMessageGroup group(getRTPSParticipant(), mp_PDPWriter, RTPSMessageGroup::WRITER, _msgbuffer);
+
+                std::vector<GUID_t> remote_readers;
+                LocatorList_t locators;
+
+                // TODO: modify announcement mechanism to allow direct message sending
+                //for (auto it = pW->matchedReadersBegin(); it != pW->matchedReadersEnd(); ++it)
+                //{
+                //    RemoteReaderAttributes & att = (*it)->m_att;
+                //    remote_readers.push_back(att.guid);
+
+                //    EndpointAttributes & ep = att.endpoint;
+                //    locators.push_back(ep.unicastLocatorList);
+                //    //locators.push_back(ep.multicastLocatorList);
+                //}
+
+                // TODO: remove when the Writer API issue is resolved
+                for (auto client : clients_)
```

Comment:
Not protected with the PDP mutex.

Reply:
agreed. I add also protection in assingRemoteEndpoints and removeRemoteEndpoints where this collection is updated.

Reply:
Possible deadlock. Remove here and take the mutex in DServerEvent.

Reply:
Answer included in above comment.